# CSCI4022 Homework 2; Review

## Due Monday, February 8 at 11:59 pm to Canvas

#### Submit this file as a .ipynb with *all cells compiled and run* to the associated dropbox.

***

Your solutions to computational questions should include any specified Python code and results as well as written commentary on your conclusions.  Remember that you are encouraged to discuss the problems with your classmates, but **you must write all code and solutions on your own**.

**NOTES**: 

- Any relevant data sets should be available on Canvas. To make life easier on the graders if they need to run your code, do not change the relative path names here. Instead, move the files around on your computer.
- If you're not familiar with typesetting math directly into Markdown then by all means, do your work on paper first and then typeset it later.  Here is a [reference guide](https://math.meta.stackexchange.com/questions/5020/mathjax-basic-tutorial-and-quick-reference) linked on Canvas on writing math in Markdown. **All** of your written commentary, justifications and mathematical work should be in Markdown.  I also recommend the [wikibook](https://en.wikibooks.org/wiki/LaTeX) for LaTex.
- Because you can technically evaluate notebook cells is a non-linear order, it's a good idea to do **Kernel $\rightarrow$ Restart & Run All** as a check before submitting your solutions.  That way if we need to run your code you will know that it will work as expected. 
- It is **bad form** to make your reader interpret numerical output from your code.  If a question asks you to compute some value from the data you should show your code output **AND** write a summary of the results in Markdown directly below your code. 
- 45 points of this assignment are in problems.  The remaining 5 are for neatness, style, and overall exposition of both code and text.
- This probably goes without saying, but... For any question that asks you to calculate something, you **must show all work and justify your answers to receive credit**. Sparse or nonexistent work will receive sparse or nonexistent credit. 
- There is *not a prescribed API* for these problems.  You may answer coding questions with whatever syntax or object typing you deem fit.  Your evaluation will primarily live in the clarity of how well you present your final results, so don't skip over any interpretations!  Your code should still be commented and readable to ensure you followed the given course algorithm.

---

In [2]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import random
import seaborn as sns

import string
import re
import unicodedata
# from cleantext import clean
# import nltk

***
<a/ id='p1'></a>
[Back to top](#top)
# Problem 1 (Theory: minhashing; 10 pts)

Consider minhash values for a single column vector that contains 10 components/rows. Seven of rows hold 0 and three hold 1. Consider taking all 10! = 3,628,800 possible distinct permutations of ten rows. When we choose a permutation of the rows and produce a minhash value for the column, we will use the number of the row, in the permuted order, that is the first with a 1.  Use Markdown cells to demonstrate answers to the following.

#### a) For exactly how many of the 3,628,800 permutations is the minhash value for the column a 9?  What proportion is this?

There would be $0$ permutations where the minhash value for the column is $9$. The proportion is also $0$. This is because there is no possible way for the 9th component to be the first $1$ to appear when there are $10$ components total with $3$ ones and $6$ zeros.

#### b) For exactly how many of the 3,628,800 permutations is the minhash value for the column a 8?

There would be $1$ permutaion where the minhash value for the column is $8$. The proportion is $\frac{1}{3,628,800}$. This is becauser there is only one possible way for the 8th component to be the first $1$ to appear when there are $10$ components total with $3$ ones and $6$ zeros. 

#### c) For exactly how many of the 3,628,800 permutations is the minhash value for the column a 3?

There would be $7!$ or $5040$ permutations where the minhash value for the column is $3$. The proporiton is $\frac{1}{720}$. This is because the first three components must be $001$ which leaves $7!$ possible permuations with this as the beggining. 

***
<a/ id='p3'></a>
[Back to top](#top)
# Problem 2 (Applied Minhashing; 35 pts)

In this problem we compare similarities of 5 documents available on http://www.gutenberg.org

 1) The first approximately 10000 characters of Miguel de Unamuno's *Niebla*, written in Spanish, in the file `niebla.txt`
 
 2) The first approximately 10000 characters of Miguel de Cervantes *The Ingenious Gentleman Don Quixote of La Mancha*, written in Spanish, in the file `DQ.txt`
 
 3) The first approximately 10000 characters of Homer's *The Odyssey*, translated into English by Samuel Butler, in the file `odyssey.txt`
 
 4) The first approximately 10000 characters of Kate Chopin's *The Awakening* in the file `awaken.txt`
 
 5) The entirety of around 12000 characters of Kate Chopin's *Beyond the Bayou* in the file `BB.txt`
 
### a) Clean the 4 documents, scrubbing all punctuation, changes cases to lower case, and removing accent marks as appropriate.  

You should have only 27 unique characters in each book/section after cleaning, corresponding to white spaces and the 26 letters.  


**For this problem, you may import any text-based packages you desire to help wrangle the data.**  I recommend looking at some functions within `string` or the RegEx `re` packages.

You can and probably should use functions in the string package such as `string.lower`, `string.replace`, etc.

All 5 documents have been saved in UTF-8 encoding.




In [22]:
filename_1 = "data/awaken.txt"
filename_2 = "data/BB.txt"
filename_3 = "data/DQ.txt"
filename_4 = "data/niebla.txt"
filename_5 = "data/odyssey.txt"

def strip_accents(s): #from this stack overflow post: https://stackoverflow.com/questions/517923/what-is-the-best-way-to-remove-accents-normalize-in-a-python-unicode-string
   return ''.join(c for c in unicodedata.normalize('NFD', s)
                  if unicodedata.category(c) != 'Mn')

#for awaken.txt
file = open(filename_1, 'rt', encoding="utf8")
text_1 = file.read()
file.close

text_1 = text_1.lower()
text_1 = re.sub(r'[^\w\s]', '', text_1) #as seen here: https://www.geeksforgeeks.org/python-remove-punctuation-from-string/
text_1 = text_1.translate({ord(k): None for k in string.digits})
text_1 = re.sub(r'\n', '', text_1)


#for BB.txt
file = open(filename_2, 'rt', encoding="utf8")
text_2 = file.read()
file.close

text_2 = text_2.lower()
text_2 = re.sub(r'[^\w\s]', '', text_2) #as seen here: https://www.geeksforgeeks.org/python-remove-punctuation-from-string/
text_2 = text_2.translate({ord(k): None for k in string.digits})
text_2 = re.sub(r'\n', '', text_2)


#for DQ.txt
file = open(filename_3, 'rt', encoding="utf8")
text_3 = file.read()
file.close

text_3 = text_3.lower()
text_3 = re.sub(r'[^\w\s]', '', text_3) #as seen here: https://www.geeksforgeeks.org/python-remove-punctuation-from-string/
text_3 = text_3.translate({ord(k): None for k in string.digits})
text_3 = strip_accents(text_3)
text_3 = re.sub(r'\n', '', text_3)


#for niebla.txt
file = open(filename_4, 'rt', encoding="utf8")
text_4 = file.read()
file.close

text_4 = text_4.lower()
text_4 = re.sub(r'[^\w\s]', '', text_4) #as seen here: https://www.geeksforgeeks.org/python-remove-punctuation-from-string/
text_4 = re.sub(r'[_]', '', text_4)
text_4 = text_4.translate({ord(k): None for k in string.digits})
text_4 = strip_accents(text_4)
text_4 = re.sub(r'\n', '', text_4)


#for odyssey.txt
file = open(filename_5, 'rt', encoding="utf8")
text_5 = file.read()
file.close

text_5 = text_5.lower()
text_5 = re.sub(r'[^\w\s]', '', text_5) #as seen here: https://www.geeksforgeeks.org/python-remove-punctuation-from-string/
text_5 = text_5.translate({ord(k): None for k in string.digits})
text_5 = re.sub(r'\n', '', text_5)

#print(text_5)


### b) Compute exact similarity scores between the documents.  Are these the expected results?

Notes:
- You may choose or explore different values of $k$ for your shingles.
- You may choose to shingle on words and create an n-gram model, but it is recommended you shingle on letters as described in class
- You may construct your characteristic matrix or characteristic sets with or without hash functions (e.g. by using `set()`).  Note that choice of hash function should change heavily with $k$!

In [51]:
# k = 1
awaken_set_1 = set(text_1)
bb_set_1 = set(text_2)
dq_set_1 = set(text_3)
niebla_set_1 = set(text_4)
odyssey_set_1 = set(text_5)

# print(len(awaken_set))
# print(len(bb_set))
# print(len(dq_set)) #missing w and k
# print(len(niebla_set)) #missing w
#print(odyssey_set_1)



# d = {'awaken': awaken_set, 'bb': bb_set, 'dq': dq_set, 'niebla': niebla_set, 'odyssey': odyssey_set}
# df = pd.DataFrame(data = d)

text_sets_1 = [awaken_set_1, bb_set_1, dq_set_1, niebla_set_1, odyssey_set_1]
# awaken_sim_1 = []

simularities_1 = np.zeros((5,5))

for i in range(len(text_sets_1)):
    for j in range(len(text_sets_1)):
        numer = len(text_sets_1[i].intersection(text_sets_1[j]))
        denom = len(text_sets_1[i].union(text_sets_1[j]))
        sim = numer/denom
        simularities_1[i][j] = sim

print("k = 1")
print(simularities_1)


#k = 2
k = 2
awaken_set_2 = set([text_1[i:i + k] for i in range(len(text_1) - k + 1)])
bb_set_2 = set([text_2[i:i + k] for i in range(len(text_2) - k + 1)])
dq_set_2 = set([text_3[i:i + k] for i in range(len(text_3) - k + 1)])
niebla_set_2 = set([text_4[i:i + k] for i in range(len(text_4) - k + 1)])
odyssey_set_2 = set([text_5[i:i + k] for i in range(len(text_5) - k + 1)])

text_sets_2 = [awaken_set_2, bb_set_2, dq_set_2, niebla_set_2, odyssey_set_2]

simularities_2 = np.zeros((5,5))

for i in range(len(text_sets_2)):
    for j in range(len(text_sets_2)):
        numer = len(text_sets_2[i].intersection(text_sets_2[j]))
        denom = len(text_sets_2[i].union(text_sets_2[j]))
        sim = numer/denom
        simularities_2[i][j] = sim

print("k = 2")
print(simularities_2)        


#k = 3
k = 3
awaken_set_3 = set([text_1[i:i + k] for i in range(len(text_1) - k + 1)])
bb_set_3 = set([text_2[i:i + k] for i in range(len(text_2) - k + 1)])
dq_set_3 = set([text_3[i:i + k] for i in range(len(text_3) - k + 1)])
niebla_set_3 = set([text_4[i:i + k] for i in range(len(text_4) - k + 1)])
odyssey_set_3 = set([text_5[i:i + k] for i in range(len(text_5) - k + 1)])

text_sets_3 = [awaken_set_3, bb_set_3, dq_set_3, niebla_set_3, odyssey_set_3]

simularities_3 = np.zeros((5,5))

for i in range(len(text_sets_3)):
    for j in range(len(text_sets_3)):
        numer = len(text_sets_3[i].intersection(text_sets_3[j]))
        denom = len(text_sets_3[i].union(text_sets_3[j]))
        sim = numer/denom
        simularities_3[i][j] = sim

print("k = 3")
print(simularities_3)


#k = 4
k = 4
awaken_set_4 = set([text_1[i:i + k] for i in range(len(text_1) - k + 1)])
bb_set_4 = set([text_2[i:i + k] for i in range(len(text_2) - k + 1)])
dq_set_4 = set([text_3[i:i + k] for i in range(len(text_3) - k + 1)])
niebla_set_4 = set([text_4[i:i + k] for i in range(len(text_4) - k + 1)])
odyssey_set_4 = set([text_5[i:i + k] for i in range(len(text_5) - k + 1)])

text_sets_4 = [awaken_set_4, bb_set_4, dq_set_4, niebla_set_4, odyssey_set_4]

simularities_4 = np.zeros((5,5))

for i in range(len(text_sets_4)):
    for j in range(len(text_sets_4)):
        numer = len(text_sets_4[i].intersection(text_sets_4[j]))
        denom = len(text_sets_4[i].union(text_sets_4[j]))
        sim = numer/denom
        simularities_4[i][j] = sim

print("k = 4")
print(simularities_4)


#k = 5
k = 5
awaken_set_5 = set([text_1[i:i + k] for i in range(len(text_1) - k + 1)])
bb_set_5 = set([text_2[i:i + k] for i in range(len(text_2) - k + 1)])
dq_set_5 = set([text_3[i:i + k] for i in range(len(text_3) - k + 1)])
niebla_set_5 = set([text_4[i:i + k] for i in range(len(text_4) - k + 1)])
odyssey_set_5 = set([text_5[i:i + k] for i in range(len(text_5) - k + 1)])

text_sets_5 = [awaken_set_5, bb_set_5, dq_set_5, niebla_set_5, odyssey_set_5]

simularities_5 = np.zeros((5,5))

for i in range(len(text_sets_5)):
    for j in range(len(text_sets_5)):
        numer = len(text_sets_5[i].intersection(text_sets_5[j]))
        denom = len(text_sets_5[i].union(text_sets_5[j]))
        sim = numer/denom
        simularities_5[i][j] = sim

print("k = 5")
print(simularities_5)  


#k = 6
k = 6
awaken_set_6 = set([text_1[i:i + k] for i in range(len(text_1) - k + 1)])
bb_set_6 = set([text_2[i:i + k] for i in range(len(text_2) - k + 1)])
dq_set_6 = set([text_3[i:i + k] for i in range(len(text_3) - k + 1)])
niebla_set_6 = set([text_4[i:i + k] for i in range(len(text_4) - k + 1)])
odyssey_set_6 = set([text_5[i:i + k] for i in range(len(text_5) - k + 1)])

text_sets_6 = [awaken_set_6, bb_set_6, dq_set_6, niebla_set_6, odyssey_set_6]

simularities_6 = np.zeros((5,5))

for i in range(len(text_sets_5)):
    for j in range(len(text_sets_5)):
        numer = len(text_sets_6[i].intersection(text_sets_6[j]))
        denom = len(text_sets_6[i].union(text_sets_6[j]))
        sim = numer/denom
        simularities_6[i][j] = sim

print("k = 6")
print(simularities_6)  


#k = 10
k = 10
awaken_set_10 = set([text_1[i:i + k] for i in range(len(text_1) - k + 1)])
bb_set_10 = set([text_2[i:i + k] for i in range(len(text_2) - k + 1)])
dq_set_10 = set([text_3[i:i + k] for i in range(len(text_3) - k + 1)])
niebla_set_10 = set([text_4[i:i + k] for i in range(len(text_4) - k + 1)])
odyssey_set_10 = set([text_5[i:i + k] for i in range(len(text_5) - k + 1)])

text_sets_10 = [awaken_set_10, bb_set_10, dq_set_10, niebla_set_10, odyssey_set_10]

simularities_10 = np.zeros((5,5))

for i in range(len(text_sets_10)):
    for j in range(len(text_sets_10)):
        numer = len(text_sets_10[i].intersection(text_sets_10[j]))
        denom = len(text_sets_10[i].union(text_sets_10[j]))
        sim = numer/denom
        simularities_10[i][j] = sim

print("k = 10")
print(simularities_10)  

k = 1
[[1.         1.         0.92592593 0.96296296 1.        ]
 [1.         1.         0.92592593 0.96296296 1.        ]
 [0.92592593 0.92592593 1.         0.96153846 0.92592593]
 [0.96296296 0.96296296 0.96153846 1.         0.96296296]
 [1.         1.         0.92592593 0.96296296 1.        ]]
k = 2
[[1.         0.8        0.56643357 0.61904762 0.75565611]
 [0.8        1.         0.55990783 0.63781321 0.7912844 ]
 [0.56643357 0.55990783 1.         0.78247734 0.54137116]
 [0.61904762 0.63781321 0.78247734 1.         0.5954023 ]
 [0.75565611 0.7912844  0.54137116 0.5954023  1.        ]]
k = 3
[[1.         0.50554734 0.28365759 0.31317886 0.47142309]
 [0.50554734 1.         0.27883539 0.3139657  0.48315868]
 [0.28365759 0.27883539 1.         0.47916667 0.26878728]
 [0.31317886 0.3139657  0.47916667 1.         0.30044346]
 [0.47142309 0.48315868 0.26878728 0.30044346 1.        ]]
k = 4
[[1.         0.28104002 0.07728803 0.09512027 0.2583423 ]
 [0.28104002 1.         0.07927179 0.09731413

### c) Implement minhashing with 1000 hash functions on the 4 documents, checking your results against those in part b).

- You may choose your own value of $p$ as the modulus of the hash functions.  You are encouraged to use the example code from the minhashing in class notebook to start you out.

In [48]:
#working from k = 5
k = 6
awaken_list_5 = [text_1[i:i + k] for i in range(len(text_1) - k + 1)]
bb_list_5 = [text_2[i:i + k] for i in range(len(text_2) - k + 1)]
dq_list_5 = [text_3[i:i + k] for i in range(len(text_3) - k + 1)]
niebla_list_5 = [text_4[i:i + k] for i in range(len(text_4) - k + 1)]
odyssey_list_5 = [text_5[i:i + k] for i in range(len(text_5) - k + 1)]

#print(awaken_list_4)
# print(len(awaken_list_4))
# print(len(bb_list_4))
# print(len(dq_list_4))
# print(len(niebla_list_4))
# print(len(odyssey_list_4))
      
dup_shingles = [*awaken_list_5, *bb_list_5, *dq_list_5, *niebla_list_5, *odyssey_list_5]

# print(len(dup_shingles))

shingles = []

[shingles.append(s) for s in dup_shingles if s not in shingles]

#print(shingles)
print(len(shingles))


awaken_sig_matrix = []
bb_sig_matrix = []
dq_sig_matrix = []
niebla_sig_matrix = []
odyssey_sig_matrix = []

for s in shingles:
    if s in text_1:
        awaken_sig_matrix.append(1)
    else:
        awaken_sig_matrix.append(0)
    
    if s in text_2:
        bb_sig_matrix.append(1)
    else:
        bb_sig_matrix.append(0)
        
    if s in text_3:
        dq_sig_matrix.append(1)
    else:
        dq_sig_matrix.append(0)
        
    if s in text_4:
        niebla_sig_matrix.append(1)
    else:
        niebla_sig_matrix.append(0)
        
    if s in text_5:
        odyssey_sig_matrix.append(1)
    else:
        odyssey_sig_matrix.append(0)
        
d = {'awaken': awaken_sig_matrix, 'bb': bb_sig_matrix, 'dq': dq_sig_matrix, 'niebla': niebla_sig_matrix, 'odyssey': odyssey_sig_matrix}
df = pd.DataFrame(data = d)
df

31131


awaken  bb  dq  niebla  odyssey
0           1   0   0       0        0
1           1   1   0       0        0
2           1   1   0       0        0
3           1   1   0       0        0
4           1   0   0       0        0
...       ...  ..  ..     ...      ...
31126       0   0   0       0        1
31127       0   0   0       0        1
31128       0   0   0       0        1
31129       0   0   0       0        1
31130       0   0   0       0        1

[31131 rows x 5 columns]

In [49]:
#taken from class nb03
def minhash(nhash, dfC):
    '''
    Takes a number of hash functions to use (nhash) and characteristic matrix (dfC)
    '''
    # use the "universal hash":  (a*x+b) mod p, where a, b are random ints and p > N (= 10 here) is prime
    np.random.seed(4022)
    Ahash = np.random.choice(range(0,10000), size=nhash)
    Bhash = np.random.choice(range(0,10000), size=nhash)
    Phash = 61051 #42042

    # STEP 2:  initialize signature matrix to all infinities

    # initialize the signature matrix
    Msig = np.full([nhash, len(dfC.columns)], fill_value=np.inf)

    # fill in the signature matrix:

    # For each row of the characteristic matrix... 
    hash_vals = [0]*nhash # initialize
    for r in range(len(dfC)):
        # STEP 3:  Compute hash values (~permuted row numbers) for that row under each hash function
        for h in range(nhash):
            hash_vals[h] = (Ahash[h]*r + Bhash[h])%Phash
        # STEP 4:  For each column, if there is a 0, do nothing...
        for c in range(len(dfC.columns)):
            # ... but if there is a 1, replace signature matrix element in that column for each hash fcn 
            # with the minimum of the hash value in this row, and the current signature matrix element
            if dfC.iloc[r,c]==1:
                for h in range(nhash):
                    if hash_vals[h] < Msig[h,c]:
                        Msig[h,c] = hash_vals[h]
    return Msig


Msig_tmp = minhash(1000, df)

In [50]:
hash_simularities = np.zeros((5,5))

for i in range(5):
    for j in range(5):
        hash_simularities[i][j] = sum(Msig_tmp[:,i]==Msig_tmp[:,j])/1000
        
print(hash_simularities)

[[1.    0.104 0.003 0.002 0.074]
 [0.104 1.    0.002 0.005 0.086]
 [0.003 0.002 1.    0.093 0.   ]
 [0.002 0.005 0.093 1.    0.003]
 [0.074 0.086 0.    0.003 1.   ]]




### d) Discussion:

Can we detect expected differences here?  Are the two Spanish docuemnts most similar to each other?  Are the two documents by the same author, with the same theme, the most similar?  What kind of alternatives might have captured the structures between these texts?



# Answer

We can detect that the English documents have relatively low simularites to the Spanish documents which was expected due to language differences. The two Spanish documents, DQ and Niebla, are more simular to eachother as opposed to the other documents and have the second highest simulurity (.093) overall after Awaken and BB being the most similar (.104) and BB and Odyssey (.086) being the third most. The two documents written by the same author, Awaken and BB, have the highest simularity between two documents at .104. A noticeable trend is that both Awaken and BB are relatively simular to Odyssey with .074 and .086 respective simularites despite being from different others which may suggest similar themes or topics. 